homeassistant/components/switch/xiaomi_ir.py

```diff
@@ -0,0 +1,181 @@
+    CONF_SWITCHES, CONF_HOST, CONF_TOKEN)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+REQUIREMENTS = ['python-miio==0.3.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+SERVICE_LEARN = 'chuangmiIr_learn_command'
+SERVICE_SEND = 'chuangmiIr_send_packet'
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_COMMAND_OFF, default=None): cv.string,
+    vol.Optional(CONF_COMMAND_ON, default=None): cv.string,
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_TOKEN): vol.All(str, vol.Length(min=32, max=32)),
+    vol.Optional(CONF_SWITCHES, default={}):
+        vol.Schema({cv.slug: SWITCH_SCHEMA}),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Xiaomi IR Remote (Chuangmi IR) platform."""
+    from miio import ChuangmiIr, DeviceException
+    from construct import ChecksumError
+
+    @asyncio.coroutine
+    def _learn_command(call):
+        """Handle a learn command."""
+        from random import randint
+        slot = randint(1, 1000000)
+
+        yield from hass.async_add_job(remote.learn, slot)
+
+        _LOGGER.info("Press the key you want Home Assistant to learn")
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=20):
+            packet = yield from hass.async_add_job(
+                remote.read, slot)
+            if packet['code']:
+                log_msg = "Recieved packet is: {}".\
+                          format(packet['code'])
+                _LOGGER.info(log_msg)
+                hass.components.persistent_notification.async_create(
+                    log_msg, title='ChuangmiIr')
+                return
+            yield from asyncio.sleep(1, loop=hass.loop)
+        _LOGGER.error("Did not received any signal")
+        hass.components.persistent_notification.async_create(
+            "Did not received any signal", title='ChuangmiIr')
+
+    @asyncio.coroutine
+    def _send_packet(call):
+        """Send a packet."""
+        packets = call.data.get('packet', [])
+        for packet in packets:
+            payload = str(packet)
+            _LOGGER.debug(payload)
+            try:
+                yield from hass.async_add_job(
+                    remote.play, payload, None)
+            except DeviceException as ex:
+                _LOGGER.error(
+                    "Failed to send packet, %s, exception: %s", payload, ex)
+
+    host = config.get(CONF_HOST)
+    token = config.get(CONF_TOKEN)
+
+    # Create handler
+    _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
+    remote = ChuangmiIr(host, token)
+
+    try:
+        remote.info()
+    except ChecksumError as ex:
+        _LOGGER.error("Token is wrong : %s", ex)
+        return
+
+    hass.services.register(DOMAIN, SERVICE_LEARN + '_' +
```

Comment:
Please apply a schema for input validation.

Reply:
> Please apply a schema for input validation.
Input is validated on line 37, this just makes the error message easier to read for people in case they use the wrong token (but token has the correct length).